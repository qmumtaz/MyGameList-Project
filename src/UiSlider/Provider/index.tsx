import React, { createContext, useCallback, useMemo, useState } from "react";

export interface ContextType {
  trackIsActive: boolean;
  setTrackIsActive: (value: boolean) => void;
  multiplier: number;
  sliderWidth: number;
  activeItem: number;
  setActiveItem: (value: number) => void;
  constraint: number;
  initSliderWidth: (width: number) => void;
  itemWidth: number;
  setMultiplier: (value: number) => void;
  setConstraint: (value: number) => void;
  setItemWidth: (value: number) => void;
  positions: number[];
  setPositions: (value: number[]) => void;
}

export const Context = createContext<ContextType | undefined>(undefined);

interface ProviderProps {
  children: React.ReactNode;
}

export const Provider: React.FC<ProviderProps> = ({ children }) => {
  const [trackIsActive, setTrackIsActive] = useState(false);
  const [multiplier, setMultiplier] = useState(0.35);
  const [sliderWidth, setSliderWidth] = useState(0);
  const [activeItem, setActiveItem] = useState(0);
  const [constraint, setConstraint] = useState(0);
  const [itemWidth, setItemWidth] = useState(0);
  const [positions, setPositions] = useState<number[]>([]);

  const initSliderWidth = useCallback(
    (width: number) => setSliderWidth(width),
    []
  );

  const value = useMemo(
    () => ({
      trackIsActive,
      setTrackIsActive,
      multiplier,
      sliderWidth,
      activeItem,
      setActiveItem,
      constraint,
      initSliderWidth,
      itemWidth,
      setMultiplier,
      setItemWidth,
      setConstraint,
      positions,
      setPositions
    }),
    [
      trackIsActive,
      setTrackIsActive,
      multiplier,
      sliderWidth,
      activeItem,
      setActiveItem,
      constraint,
      initSliderWidth,
      itemWidth,
      setMultiplier,
      setItemWidth,
      setConstraint,
      positions,
      setPositions
    ]
  );

  return <Context.Provider value={value}>{children}</Context.Provider>;
};
