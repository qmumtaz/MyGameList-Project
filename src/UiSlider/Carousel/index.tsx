import { useMediaQuery, useTheme } from "@chakra-ui/react";
import React, { useContext, useEffect } from "react";
import Item from "../Item";
import { Context, ContextType } from "../Provider";
import Slider from "../Slider";
import Track from "../Track";

export interface CarouselPropTypes {
  children: React.ReactNode[];
  gap: number;
}

export const Carousel: React.FC<CarouselPropTypes> = ({ children, gap }) => {
  const context = useContext(Context);

  const {
    setItemWidth,
    sliderWidth,
    setMultiplier,
    setConstraint,
    itemWidth,
    setPositions,
  } = context as ContextType;

  const { breakpoints } = useTheme();

  useEffect(() => {
    const newPositions = children?.map(
      (_, index) => -Math.abs((itemWidth + gap) * index)
    );

    setPositions(newPositions);
  }, [children, gap, itemWidth, setPositions]);

  const [isBetweenBaseAndMd] = useMediaQuery(
    `(min-width: ${breakpoints?.base}) and (max-width: ${breakpoints?.md})`
  );

  const [isBetweenMdAndXl] = useMediaQuery(
    `(min-width: ${breakpoints?.md}) and (max-width: ${breakpoints?.xl})`
  );

  const [isGreaterThanXL] = useMediaQuery(`(min-width: ${breakpoints?.xl})`);

  useEffect(() => {
    if (isBetweenBaseAndMd) {
      setItemWidth(sliderWidth - gap);
      setMultiplier(0.65);
      setConstraint(1);
    }
    if (isBetweenMdAndXl) {
      setItemWidth(sliderWidth / 2 - gap);
      setMultiplier(0.5);
      setConstraint(2);
    }
    if (isGreaterThanXL) {
      setItemWidth(sliderWidth / 3 - gap);
      setMultiplier(0.35);
      setConstraint(3);
    }
  }, [
    isBetweenBaseAndMd,
    isBetweenMdAndXl,
    isGreaterThanXL,
    sliderWidth,
    gap,
    setItemWidth,
    setMultiplier,
    setConstraint,
  ]);

  return (
    <Slider gap={gap}>
      <Track>
        {children.map((child, index) => (
          <Item gap={gap} key={index} index={index}>
            {child}
          </Item>
        ))}
      </Track>
    </Slider>
  );
};
