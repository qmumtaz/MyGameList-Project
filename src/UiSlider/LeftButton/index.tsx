import {
  Box,
  Button,
  ButtonProps,
  ComponentWithAs,
  IconProps,
} from "@chakra-ui/react";
import { ChevronLeftIcon } from "@chakra-ui/icons";
import React, { useContext } from "react";
import { Context, ContextType } from "../Provider";

interface CustomButtonProps extends ButtonProps {
  customIcon?: ComponentWithAs<"svg", IconProps> | React.Element;
}

export const LeftButton = ({
  customIcon,
  ...props
}: CustomButtonProps ) => {
  const context = useContext(Context);

  const { setTrackIsActive, activeItem, positions, setActiveItem } =
    context as ContextType;

  const handleFocus = () => setTrackIsActive(true);

  const handleDecrementClick = () => {
    setTrackIsActive(true);
    !(activeItem === positions.length - positions.length) &&
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      setActiveItem((prev: number) => prev - 1);
  };
  return (
    <>
      <Button
        {...props}
        onClick={handleDecrementClick}
        onFocus={handleFocus}
        zIndex={2}
        minW={0}
      >
        {customIcon ? <Box>{customIcon}</Box> : <ChevronLeftIcon boxSize={8} />}
      </Button>
    </>
  );
};
