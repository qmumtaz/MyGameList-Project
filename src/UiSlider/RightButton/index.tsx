import { ChevronRightIcon } from "@chakra-ui/icons";
import {
  Box,
  Button,
  ButtonProps,
  ComponentWithAs,
  IconProps,
} from "@chakra-ui/react";
import React, { useContext } from "react";
import { Context, ContextType } from "../Provider";

interface CustomButtonProps extends ButtonProps {
  customIcon?: ComponentWithAs<"svg", IconProps> | React.Element;
}

export const RightButton = ({ customIcon, ...props }: CustomButtonProps) => {
  const context = useContext(Context);

  const { setTrackIsActive, activeItem, constraint, positions, setActiveItem } =
    context as ContextType;

  const handleFocus = () => setTrackIsActive(true);

  const handleIncrementClick = () => {
    setTrackIsActive(true);
    !(activeItem === positions.length - constraint) &&
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      setActiveItem((prev: number) => prev + 1);
  };
  return (
    <>
      <Button
        {...props}
        onClick={handleIncrementClick}
        onFocus={handleFocus}
        zIndex={2}
        minW={0}
      >
        {customIcon ? (
          <Box>{customIcon}</Box>
        ) : (
          <ChevronRightIcon boxSize={8} />
        )}
      </Button>
    </>
  );
};
