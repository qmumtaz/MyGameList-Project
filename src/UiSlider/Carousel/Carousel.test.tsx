import React from "react";
import { Carousel } from ".";
import renderer from "react-test-renderer";
import { DemoComp } from "./Carousel.stories";
import { Provider } from "../Provider";
import { ChakraProvider } from "@chakra-ui/react";
import { LeftButton } from "../LeftButton";
import { RightButton } from "../RightButton";

describe("renders correctly", () => {
  it("Carousel renders correctly", () => {
    const container = renderer
      .create(
        <ChakraProvider>
          <Provider>
            <Carousel gap={3}>
              <DemoComp index={1} />
              <DemoComp index={2} />
              <DemoComp index={3} />
              <DemoComp index={4} />
            </Carousel>
          </Provider>
        </ChakraProvider>
      )
      .toJSON();
    expect(container).toMatchSnapshot();
  });
});

describe("active item changes successfully", () => {
  xit('active index increases by 1 when "next" button is clicked', () => {
    const container = renderer.create(
      <Provider>
        <Carousel gap={50}>
          <DemoComp index={1} />
          <DemoComp index={2} />
          <DemoComp index={3} />
          <DemoComp index={4} />
          <DemoComp index={5} />
          <DemoComp index={6} />
          <DemoComp index={7} />
          <DemoComp index={8} />
          <DemoComp index={9} />
          <DemoComp index={10} />
        </Carousel>
        <LeftButton />
        <RightButton />
      </Provider>
    ).toJSON();
    expect(container).toMatchSnapshot();
  });
});
