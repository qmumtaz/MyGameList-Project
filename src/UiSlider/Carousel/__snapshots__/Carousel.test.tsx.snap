// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Carousel renders correctly 1`] = `
Array [
  <div
    className="css-seh9k4"
  >
    0
  </div>,
  <span />,
]
`;
